#! /bin/sh /usr/share/dpatch/dpatch-run
## 09_runasuser.dpatch by  <mpitt@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Provide a configuration switch --enable-privilege-dropping for running
## DP: the scheduler as a normal system user. This confines security
## DP: vulnerabilities to the scheduler itself rather than providing a
## DP: remote root attack vector.
## DP: Forwarded upstream: http://www.cups.org/str.php?L2370

@DPATCH@
diff -urNad cups-1.2-ubuntu~/config-scripts/cups-defaults.m4 cups-1.2-ubuntu/config-scripts/cups-defaults.m4
--- cups-1.2-ubuntu~/config-scripts/cups-defaults.m4	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/config-scripts/cups-defaults.m4	2007-04-30 11:47:55.000000000 +0200
@@ -220,6 +220,17 @@
 AC_DEFINE_UNQUOTED(CUPS_DEFAULT_GROUP, "$CUPS_GROUP")
 AC_DEFINE_UNQUOTED(CUPS_DEFAULT_SYSTEM_GROUPS, "$CUPS_SYSTEM_GROUPS")
 
+dnl Privilege dropping
+AC_ARG_ENABLE(privilege-dropping, [  --enable-privilege-dropping       drop root privileges to normal user, default=no])
+if test "x$enable_privilege_dropping" = xyes; then
+	CUPS_DROP_PRIVILEGES=1
+	AC_DEFINE_UNQUOTED(CUPS_DROP_PRIVILEGES, 1)
+else
+	CUPS_DROP_PRIVILEGES=0
+	AC_DEFINE_UNQUOTED(CUPS_DROP_PRIVILEGES, 0)
+fi
+AC_SUBST(CUPS_DROP_PRIVILEGES)
+
 dnl Default printcap file...
 AC_ARG_WITH(printcap, [  --with-printcap         set default printcap file],
 	default_printcap="$withval",
diff -urNad cups-1.2-ubuntu~/config.h.in cups-1.2-ubuntu/config.h.in
--- cups-1.2-ubuntu~/config.h.in	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/config.h.in	2007-04-30 11:47:55.000000000 +0200
@@ -41,6 +41,11 @@
 #define CUPS_DEFAULT_GROUP	"sys"
 #define CUPS_DEFAULT_SYSTEM_GROUPS	"sys root system"
 
+/*
+ * Privilege dropping
+ */
+
+#define CUPS_DROP_PRIVILEGES	0
 
 /*
  * Default file permissions...
diff -urNad cups-1.2-ubuntu~/scheduler/cert.c cups-1.2-ubuntu/scheduler/cert.c
--- cups-1.2-ubuntu~/scheduler/cert.c	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/scheduler/cert.c	2007-04-30 11:47:55.000000000 +0200
@@ -116,7 +116,7 @@
     * Root certificate...
     */
 
-    fchmod(fd, 0440);
+    fchmod(fd, 0240);
     fchown(fd, RunUser, SystemGroupIDs[0]);
 
     cupsdLogMessage(CUPSD_LOG_DEBUG2, "cupsdAddCert: NumSystemGroups=%d",
diff -urNad cups-1.2-ubuntu~/scheduler/conf.c cups-1.2-ubuntu/scheduler/conf.c
--- cups-1.2-ubuntu~/scheduler/conf.c	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/scheduler/conf.c	2007-04-30 11:47:55.000000000 +0200
@@ -465,7 +465,11 @@
   if (!status)
     return (0);
 
+#if CUPS_DROP_PRIVILEGES == 1
+  RunUser = User;
+#else
   RunUser = getuid();
+#endif
 
  /*
   * See if the ServerName is an IP address...
diff -urNad cups-1.2-ubuntu~/scheduler/cups-polld.c cups-1.2-ubuntu/scheduler/cups-polld.c
--- cups-1.2-ubuntu~/scheduler/cups-polld.c	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/scheduler/cups-polld.c	2007-04-30 11:47:55.000000000 +0200
@@ -34,6 +34,9 @@
  * Include necessary headers...
  */
 
+#include "config.h"
+#include <pwd.h>
+
 #include <cups/http-private.h>
 #include <cups/cups.h>
 #include <stdlib.h>
@@ -150,6 +153,19 @@
   }
 
  /*
+  * Drop our privileges to the cupsd scheduler user
+  */
+#if CUPS_DROP_PRIVILEGES == 1
+  struct passwd * pwd = getpwnam(CUPS_DEFAULT_USER);
+  if (!pwd) {
+    fprintf(stderr, "ERROR: could not get passwd data for user " CUPS_DEFAULT_USER "\n");
+    exit(1);
+  }
+  setgid(pwd->pw_gid);
+  setuid(pwd->pw_uid);
+#endif
+
+ /*
   * Loop forever, asking for available printers and classes...
   */
 
diff -urNad cups-1.2-ubuntu~/scheduler/main.c cups-1.2-ubuntu/scheduler/main.c
--- cups-1.2-ubuntu~/scheduler/main.c	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/scheduler/main.c	2007-04-30 11:47:55.000000000 +0200
@@ -58,6 +58,9 @@
 #include <sys/resource.h>
 #include <syslog.h>
 #include <grp.h>
+#if CUPS_DROP_PRIVILEGES == 1
+#include <pwd.h>
+#endif
 
 #ifdef HAVE_LAUNCH_H
 #  include <launch.h>
@@ -536,6 +539,20 @@
   cupsdStartSystemMonitor();
 #endif /* __APPLE__ */
 
+#if CUPS_DROP_PRIVILEGES == 1
+  /*
+   * Drop root privileges
+   */
+   struct passwd * pwd = getpwuid(User);
+   if (!pwd) {
+     cupsdLogMessage(CUPSD_LOG_ERROR, "could not get passwd data for user id %i", User);
+     exit(EXIT_FAILURE);
+   }
+   setgid(Group);
+   initgroups(pwd->pw_name, pwd->pw_gid);
+   setuid(User);
+#endif
+
  /*
   * Start any pending print jobs...
   */
@@ -1037,7 +1054,7 @@
     */
 
     if ((current_time - RootCertTime) >= RootCertDuration && RootCertDuration &&
-        !RunUser && cupsArrayCount(Clients))
+        cupsArrayCount(Clients))
     {
      /*
       * Update the root certificate...
diff -urNad cups-1.2-ubuntu~/scheduler/process.c cups-1.2-ubuntu/scheduler/process.c
--- cups-1.2-ubuntu~/scheduler/process.c	2007-04-30 11:47:51.000000000 +0200
+++ cups-1.2-ubuntu/scheduler/process.c	2007-04-30 11:48:06.000000000 +0200
@@ -245,6 +245,7 @@
       if (setuid(User))
         exit(errno);
     }
+#if CUPS_DROP_PRIVILEGES == 0
     else
     {
      /*
@@ -254,6 +255,7 @@
       setgid(Group);
       setgroups(1, &Group);
     }
+#endif
 
    /*
     * Change umask to restrict permissions on created files...
